Try: fix Safari flickering issues.

jasmussen opened the pull request:
## Description

Hopefully/maybe fixes #30803.

Safari has trouble rendering elements that use CSS3 transitions. According to basic testing in this branch, and to [this document](https://coderwall.com/p/xbjtbg/fix-element-flickering-in-safari-caused-by-css3-transitions), the attached hack should fix it.

Note that if this works, we should consider scoping the hack to just the blocks causing the issue, or somehow otherwise limiting it as best we can. While it _seemed_ harmless in my testing, as far as I understand this changes the fundamental rendering method used for blocks, and if past experience tells us anything, it is that it a) makes it way more performant, and b) sometimes interferes with absolute or sticky positioning, especially in cases where you are intentionally applying `position: static;` to a parent to inherit from an ancestor itself — this property _might_ effectively apply `position: relative;`.

That's also a note to test #30047 with this.

## How has this been tested?

Please test various blocks in the editor, in Safari. Select blocks, scroll, and verify that nothing flickers.

Blocks that commonly cause flickering appear to be the Gallery block, but sometimes also the pullquote, it's a bit unclear and also seems to depend on amount of blocks on the page. Using the demo content might be a good way to start.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Try: fix Safari flickering issues.

ellatrix commented on packages/block-library/src/reset.scss:
```diff
@@ -126,4 +126,15 @@
 		transform: revert;
 	}
+
+	// Safari can have renderint issues with CSS3 transitions.
```
What is renderint? And which CSS3 transition is causing this? The block moving one?

jasmussen replied:
Sorry, typo, "rendering".

I'm not entirely sure which property is causing the issue, and I was hoping that by starting this PR (since it seems to fix the issue for me) as a good way get the conversation going.

In my testing, the flickering happens mostly on the classic gallery buttons (left/right/remove), while you are scrolling:

![gallery](https://user-images.githubusercontent.com/1204802/116855813-4d2e8480-abfa-11eb-8018-1cf20d7e933b.gif)

But according to the initial ticket, it appears to be happening to other blocks as well.

jasmussen committed:
Fix typo.

kjellr commented:
I'm still seeing some flicker/rendering issues, using [test content](https://gist.github.com/kjellr/87f7f8168e107479a13b0a0c94d8f6b8) similar to that supplied in #30803. 😞 


https://user-images.githubusercontent.com/1202812/116884100-587eb180-abf4-11eb-8421-984cd0ca5170.mp4

jasmussen commented:
Thank you for the test content. Just to be sure, you're referring here to the fixed position background of the cover block, correct? The choppiness?

![fixed](https://user-images.githubusercontent.com/1204802/116886001-d957a080-ac28-11eb-8ff8-b3c82e8b776d.gif)

I see that, and will investigate. But it seems like an additional issue to some of the more jarring flickering I see in videos on https://github.com/WordPress/gutenberg/issues/30803, does that seem right? As in, the text popping in and out of existing, or the toolbar blinking, that seems gone in this branch on my end.

jasmussen commented:
Specifically, it appears Safari has relatively poor support for `background-attachment: fixed`, as discussed in #17718. 

@kjell if you disable the fixed background, do you still see flickering? Because I think we've got those two separate issues going on, and the fixed background is apparently an old issue with Safari. One we should fix, absolutely, but maybe separately.

kjellr commented:
@jasmussen Good news! I tested again with [the expanded test content in this comment](https://github.com/WordPress/gutenberg/issues/30803#issuecomment-827757930) (but with the background un-`fixed`), and I definitely did see an improvement with this PR active. 

I do still see the background glitching from before when the cover background is `fixed`, so I think you're right — that may be a separate issue. In the meantime, what you have here does improve the flickering in my testing at least. 

Trunk:

https://user-images.githubusercontent.com/1202812/116900505-cf24aa80-ac06-11eb-82a9-c07030cc2ec5.mp4


This PR: 

https://user-images.githubusercontent.com/1202812/116900526-d350c800-ac06-11eb-8941-186214aa4ac3.mp4

jasmussen commented:
Yeah I do think the background glitching is something we need to find a solution for. I feel like the fact that the Cover block was recently refactored to use an Image element in most cases might let us use `position: fixed;` perhaps combined with some `object-fit`, to accomplish the same as `background-attachment: fixed;`, but without any of the downsides. I'll put that on my ToDo, but in the mean time, it'd be nice to get some reviews so we can find out how disruptive this hack might be. Thanks Kjell!

kjellr commented on packages/block-library/src/reset.scss:
```diff
@@ -126,4 +126,15 @@
+		// by keeping the prefixes we're reminded that this is a hack.
+		/* eslint-disable property-no-vendor-prefix */
+		-webkit-perspective: 1;
+		-webkit-backface-visibility: none;
```
@jasmussen in my testing, this doesn't even get picked up. Are you sure it's necessary? 

<img width="300" alt="Screen Shot 2021-05-04 at 8 12 58 AM" src="https://user-images.githubusercontent.com/1202812/117001834-d7d1bb00-acb0-11eb-9a0d-f2245baa46e3.png">

kjellr replied:
It looks like `none` isn't [a supported value](https://css-tricks.com/almanac/properties/b/backface-visibility/#values) here. But this PR seems to work fine without this rule anyway, so maybe we can just remove it?

jasmussen replied:
Wild, so it seems like just adding that perspective is enough. I would think it has the same effect as adjacent properties — telling the browser to render it on the GPU — but from [the initial piece I read about it](https://coderwall.com/p/xbjtbg/fix-element-flickering-in-safari-caused-by-css3-transitions) suggested the backface property was needed as well. But yeah, if it works, we can remove it.

eduard-co commented:
How is this coming along? The flickering is driving me nuts and I wish to keep using Safari for editing my websites.

jasmussen commented:
I'm happy to report that I think I found a fix in https://github.com/WordPress/gutenberg/pull/32188. Please test that if you can, and I will close this one in favor of that.
